README.md
# Circonus Log Watch

A small utility for extracting metrics from log files and forwarding to Circonus.

> Note: In v1 all metrics use stream tags, for non-stream tagged metrics use the v0 branch/releases.

## Install

1. `mkdir -p /opt/circonus/{sbin,etc,etc/log.d}`
1. Download [latest release](../../releases/latest) from repository (or [build manually](#manual-build))
1. Extract archive into `/opt/circonus`
1. Create config in `/opt/circonus/etc` and log configs in `/opt/circonus/etc/log.d`

## Options

```sh
/opt/circonus/sbin/circonus-logwatchd -h

Flags:
      --api-app string              [ENV: CLW_API_APP] Circonus API Token app (default "circonus-logwatch")
      --api-ca-file string          [ENV: CLW_API_CA_FILE] Circonus API CA certificate file (optional, Inside API server not using public certs)
      --api-key string              [ENV: CLW_API_KEY] Circonus API Token key or 'cosi' to use COSI config
      --api-url string              [ENV: CLW_API_URL] Circonus API URL (default "https://api.circonus.com/v2/")
  -c, --config string               config file (default is /opt/circonus/etc/circonus-logwatch.(json|toml|yaml)
  -d, --debug                       [ENV: CLW_DEBUG] Enable debug messages
      --debug-cgm                   [ENV: CLW_DEBUG_CGM] Enable CGM & API debug messages
      --debug-metric                [ENV: CLW_DEBUG_METRIC] Enable metric rule evaluation tracing debug messages
      --debug-tail                  [ENV: CLW_DEBUG_TAIL] Enable log tailing messages
      --dest string                 [ENV: CLW_DESTINATION] Destination[agent|check|log|statsd] type for metrics (default "log")
      --dest-agent-interval string  [ENV: CLW_DEST_AGENT_INTERVAL] Destination[agent] Interval for metric submission to agent (default "60s")
      --dest-cid string             [ENV: CLW_DEST_CID] Destination[check] Check ID (not check bundle)
      --dest-id string              [ENV: CLW_DEST_ID] Destination[statsd|agent] metric group ID (default "circonus-logwatch")
      --dest-instance-id string     [ENV: CLW_DEST_INSTANCE_ID] Destination[check] Check Instance ID
      --dest-port string            [ENV: CLW_DEST_PORT] Destination[agent|statsd] port (agent=2609, statsd=8125)
      --dest-statsd-prefix string   [ENV: CLW_DEST_STATSD_PREFIX] Destination[statsd] Prefix prepended to every metric sent to StatsD (default "host.")
      --dest-tag string             [ENV: CLW_DEST_TAG] Destination[check] Check search tag
      --dest-target string          [ENV: CLW_DEST_TARGET] Destination[check] Check target (default hostname)
      --dest-url string             [ENV: CLW_DEST_URL] Destination[check] Check Submission URL
  -h, --help                        help for circonus-logwatch
  -l, --log-conf-dir string         [ENV: CLW_PLUGIN_DIR] Log configuration directory (default "/opt/circonus/etc/log.d")
      --log-level string            [ENV: CLW_LOG_LEVEL] Log level [(panic|fatal|error|warn|info|debug|disabled)] (default "info")
      --log-pretty                  [ENV: CLW_LOG_PRETTY] Output formatted/colored log lines
      --show-config                 Show config (json|toml|yaml) and exit
      --stat-port string            [ENV: CLW_STAT_PORT] Exposes app stats while running (default "33284")
  -V, --version                     Show version and exit

```

## Destinations

* `--dest check` metrics are sent directly to the circonus broker (will create a check if `--dest-cid` not provided). `--dest-instance-id`, `--dest-target`, and `--dest-tag` can be used to customize the check created.
* `--dest agent` metrics are sent to `/write` endpoint of local circonus-agent (`http://localhost:2609/write/id`) uses `--dest-id` to categorize the metrics. `--dest-port` controls the agent port (default 2609)
* `--dest statsd` metrics sent to statsd listener of local circonus-agent (`localhost:8125`) uses `--statsd-prefix` for each metric name, followed by `--dest-id` (`--dest-statsd-prefix` should match circonus-agent `--statsd-host-prefix` to ensure metrics are routed to correct destination by the agent). `--dest-port` controls the agent statsd port (default 8125)

## Config

Create a JSON, YAML, or TOML config in `/opt/circonus/etc/circonus-logwatch.(json|yaml|toml)`. Or, use environment variables and/or command line parameters.

YAML with a StatsD destination (send metrics to local circonus-agent statsd listener)

```yaml
---
app_stat_port: "33284"
debug: false
debug_cgm: false
debug_metric: false
debug_tail: false
log_conf_dir: /opt/circonus/etc/log.d
api:
  key: circonus api token key
  app: circonus-logwatch
  url: 'https://api.circonus.com/v2/'
  ca_file: path to api ca file, if needed
destination:
  type: statsd
  config:
    id: circonus-logwatch
    port: "8125"
    statsd_prefix: "host."
```

JSON with a Check destination (send metrics directly to a Circonus check, a check will be created if not supplied via cid or url config settings)

> NOTE: use `cid` or `url` to identify an existing check. Or, to search for a check a combination of `target`, `search_tag` and/or `instance_id`.

```json
{
    "api": {
        "app": "circonus-logwatch",
        "ca_file": "path to api ca file, if needed",
        "key": "circonus api token key",
        "url": "https://api.circonus.com/v2/"
    },
    "app_stat_port": "33284",
    "debug": false,
    "debug_cgm": false,
    "debug_metric": false,
    "debug_tail": false,
    "destination": {
        "type": "check",
        "config": {
            "cid": "check id (of existing check)",
            "url": "submission url of existing check",
            "target": "to find|create a check",
            "search_tag": "to find|create a check",
            "instance_id": "to find|create a check"
        }
    },
    "log": {
        "level": "info",
        "pretty": false
    },
    "log_conf_dir": "/opt/circonus/etc/log.d"
}
```

TOML with an Agent destination (send metrics to local circonus-agent)

```toml
app_stat_port = "33284"
debug = false
debug_cgm = false
debug_metric = false
debug_tail = false
log_conf_dir = "/opt/circonus/etc/log.d"

[api]
  key = "circonus api token key"
  app = "circonus-logwatch"
  url = "https://api.circonus.com/v2/"
  ca_file = "path to api ca file, if needed"

[destination]
  type = "agent"

  [destination.config]
    id = "circonus-logwatch"
    port = "2609"

[log]
  level = "info"
  pretty = false
```

## Log configs

Create one config (JSON, YAML, or TOML) in `--log-conf-dir` for each distinct log. Examples [`etc/log.d`](etc/log.d/) in this repository

1. `id` of the log, short identifier - optional, the base file name will be used if omitted
1. `log_file` path to the log
1. `metrics` a list of:
    1. `match` regular expression to identify lines and optionally extract named subexpressions for value and metric name
    1. `name` a static string to use as the metric name or a template for naming the metric if named subexpressions were used in match regex
    1. `tags` comma separated list of k:v pairs, templating can be used accessing named subexpressions (e.g. `foo:bar,yabba:dabba` or `foo:{{.id}},bar:baz`)
    1. `type` what type of metric (all numbers are 64bit)
        * `c` counter int
        * `g` gauge int or float
        * `ms` timing (treated as histogram) value can be a float (3.12) or duration (25ms, 3.2s, 1m10s, etc.) durations are converted to milliseconds, floats are assumed to already represent milliseconds
        * `h` histogram float
        * `s` set (ala statsd set metrics) unique string to count
        * `t` text string

### Log configuration notes

* any metric which does not have a `type` will be treated as a counter.
* any metric which does not have a subexpression named '*Value*' (case insensitive) will be treated as a counter.
* named subexpressions can be used in the name template and tag list with the following syntax `{{.id}}` where `id` is the name given to a named subexpression in the match regex
* metrics will have a stream tag added for the log `id` (e.g. for a log with an id of "foo" the tag would be `log_id:foo`)

## Manual build

1. Clone repo (outside if `GOPATH`)`git clone https://github.com/circonus-labs/circonus-logwatch && cd circonus-logwatch`
1. Build `go build -o circonus-logwatchd`
1. Ensure target directories exist `mkdir -p /opt/circonus/logwatch/{sbin,etc,etc/log.d}`
1. Install `cp circonus-logwatchd /opt/circonus/logwatch/sbin`


CHANGELOG.md
# **unreleased**

## v1.1.2

* build(deps): bump github.com/spf13/viper from 1.18.1 to 1.18.2
* build(deps): bump github.com/spf13/viper from 1.18.0 to 1.18.1
* build(deps): bump github.com/spf13/viper from 1.17.0 to 1.18.0
* build(deps): bump actions/setup-go from 4 to 5

## v1.1.1

* build: add after hook for `grype` on generated sboms
* build: add .sbom for archive artifacts
* build: update before hooks for `go mod tidy`, `govulncheck` and `golangci-lint`
* build(deps): bump golang.org/x/sys from 0.14.0 to 0.15.0

## v1.1.0

* fix: add ReadHeaderTimeout to stats server
* fix(lint): typos, unused args, struct alignment
* fix: update to current syntax (release)
* chore: remove deprecated linters
* build(deps): bump github.com/nxadm/tail from 1.4.8 to 1.4.11
* build(deps): bump github.com/circonus-labs/circonus-gometrics/v3 from 3.4.6 to 3.4.7
* build(deps): bump golang.org/x/sys from 0.12.0 to 0.14.0
* build(deps): bump github.com/rs/zerolog from 1.29.1 to 1.31.0
* build(deps): bump golang.org/x/sync from 0.3.0 to 0.5.0
* build(deps): bump github.com/spf13/viper from 1.10.1 to 1.17.0
* build(deps): bump github.com/spf13/cobra from 1.3.0 to 1.8.0
* build(deps): bump github.com/pelletier/go-toml from 1.9.4 to 1.9.5
* fix: update to latest in lint workflow

## v1.0.3

* fix: typo in systemd service file

## v1.0.2

* fix: config unit tests
* fix: test tcp port for agent config
* fix: lowercase config err msgs

## v1.0.1

* fix: lint issues
* upd: remove pkg/errors dependency
* add: lint config
* upd: all dependencies

## v1.0.0

* upd: clarify processing metric vs sending log message
* upd: go1.15
* upd: cgm and viper dependencies
* fix: exec path in service config, add `sbin`

## v1.0.0-beta.3

* fix: let cgm handle retrying check/agent destinations rather than testing the port
* upd: version information in release package
* upd: skip README.md in `log.d`

## v1.0.0-beta.2

* fix: do not skip config if log file not currently available let tail wait for the log
* fix: do not exit if only one config with a missing log file
* upd: switch to nxadm/tail updated tail (hpcloud/tail is not currently being updated/maintained)
* upd: switch to tail poll for better handling of deleted log files

## v1.0.0-beta.1

* UPDATE - stream tag only release
* add: stream tag support
* upd: log id in tag `log_id`
* add: support `tags` attribute to metric stanza in configuration
* add: support interpolation of named regex patterns in `tags` attribute
* upd: dependencies (cgm)

## v0.6.0

* Merge pull request #12 from yargevad/agent-interval
* upd: dependencies, tidy
* upd: deprecated syntax
* add: agent interval to config, command line
* fix: lint issues
* add: build linting
* doc: agent interval

## v0.5.4

* fix: clarify setting for check, it is a check id not a check bundle id

## v0.5.3

* upd: dependencies (cgm - allow/deny checks default, etc.)

## v0.5.2

* upd: dependencies (yaml, x/sync)

## v0.5.1

* upd: dependencies
* upd: disable dir/file permissions tests
* add: ctx cancel function for Stop to use
* fix: calls to New in tests to include a parent context

## v0.5.0

* upd: include service and example configs in release
* add: systemd service configuration in `service/`
* add: example config `etc/example-circonus-logwatch.yaml`
* upd: change found metric message priority from Info to Debug
* upd: tomb->context+errgroup
* add: api ca file load/config for circonus metric destination
* upd: switch to circonus-gometrics v3
* upd: condense/consolidate code
* upd: switch to go mod

## v0.4.0

* upd: release file names use x86_64, facilitate automated builds and testing
* upd: goreleaser, turn off draft

## v0.3.1

* add: freebsd to release

## v0.3.0

* add: example of apache access log  latency histograms
* doc: include `etc/log.d` in release by adding a stub readme
* fix: correctly error when only one config with an error in it
* doc: formatting

## v0.2.0

* adjust handling of tail behavior for rotated logs and nil lines
* fix: statsd base metric name, add separator
* check destination, add api config

## v0.1.1

* update readme, clarify metric type descriptions
* common appstats package

## v0.1.0

* Initial development release


cmd/root.go
// Copyright © 2017 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

package cmd

import (
	"fmt"
	stdlog "log"
	"os"
	"time"

	"github.com/circonus-labs/circonus-logwatch/internal/agent"
	"github.com/circonus-labs/circonus-logwatch/internal/config"
	"github.com/circonus-labs/circonus-logwatch/internal/config/defaults"
	"github.com/circonus-labs/circonus-logwatch/internal/release"
	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var cfgFile string

// RootCmd represents the base command when called without any subcommands.
var RootCmd = &cobra.Command{
	Use:   release.NAME,
	Short: "A small utility to send metrics extracted from log files to Circonus.",
	Long: `Send metrics extracted from log files to Circonus.

Not all useful metrics can be sent directly to a centralized system for analysis
and alerting. Often, there are valuable metrics sequestered in system and application
log files. These logs are not always in a common, easily parsable format.

Using named regular expressions and templates, this utility offers the ability
to extract these metrics from the logs and send them directly to the Circonus
system using one of several methods.

See https://github.com/circonus-labs/circonus-logwatch/etc for example configurations.
`,
	PersistentPreRunE: initLogging,
	Run: func(cmd *cobra.Command, args []string) {
		//
		// show version and exit
		//
		if viper.GetBool(config.KeyShowVersion) {
			fmt.Printf("%s v%s - commit: %s, date: %s, tag: %s\n", release.NAME, release.VERSION, release.COMMIT, release.DATE, release.TAG)
			return
		}

		//
		// show configuration and exit
		//
		if viper.GetString(config.KeyShowConfig) != "" {
			if err := config.ShowConfig(os.Stdout); err != nil {
				log.Fatal().Err(err).Msg("show-config")
			}
			return
		}

		log.Info().
			Int("pid", os.Getpid()).
			Str("name", release.NAME).
			Str("ver", release.VERSION).Msg("Starting")

		a, err := agent.New()
		if err != nil {
			log.Fatal().Err(err).Msg("Initializing")
		}

		_ = config.StatConfig()

		if err := a.Start(); err != nil {
			log.Fatal().Err(err).Msg("Startup")
		}

	},
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := RootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func bindFlagError(flag string, err error) {
	if err != nil {
		log.Fatal().Err(err).Str("flag", flag).Msg("binding flag")
	}
}
func bindEnvError(envVar string, err error) {
	if err != nil {
		log.Fatal().Err(err).Str("var", envVar).Msg("binding env var")
	}
}

func init() {
	zerolog.TimeFieldFormat = time.RFC3339Nano
	zerolog.SetGlobalLevel(zerolog.InfoLevel)
	zlog := zerolog.New(zerolog.SyncWriter(os.Stderr)).With().Timestamp().Logger()
	log.Logger = zlog

	stdlog.SetFlags(0)
	stdlog.SetOutput(zlog)

	cobra.OnInitialize(initConfig)

	desc := func(desc, env string) string {
		return fmt.Sprintf("[ENV: %s] %s", env, desc)
	}

	//
	// Basic
	//
	{
		var (
			longOpt     = "config"
			shortOpt    = "c"
			description = "config file (default is " + defaults.EtcPath + "/" + release.NAME + ".(json|toml|yaml)"
		)
		RootCmd.PersistentFlags().StringVarP(&cfgFile, longOpt, shortOpt, "", description)
	}

	{
		const (
			key         = config.KeyLogConfDir
			longOpt     = "log-conf-dir"
			shortOpt    = "l"
			envVar      = release.ENVPREFIX + "_PLUGIN_DIR"
			description = "Log configuration directory"
		)

		RootCmd.Flags().StringP(longOpt, shortOpt, defaults.LogConfPath, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.LogConfPath)
	}

	//
	// Destination for metrics
	//
	{
		const (
			key         = config.KeyDestType
			longOpt     = "dest"
			envVar      = release.ENVPREFIX + "_DESTINATION"
			description = "Destination[agent|check|log|statsd] type for metrics"
		)

		RootCmd.Flags().String(longOpt, defaults.DestinationType, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.DestinationType)
	}
	{
		const (
			key         = config.KeyDestCfgID
			longOpt     = "dest-id"
			envVar      = release.ENVPREFIX + "_DEST_ID"
			description = "Destination[statsd|agent] metric group ID"
		)

		RootCmd.Flags().String(longOpt, release.NAME, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgCID
			longOpt     = "dest-cid"
			envVar      = release.ENVPREFIX + "_DEST_CID"
			description = "Destination[check] Check ID (not check bundle id)"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgURL
			longOpt     = "dest-url"
			envVar      = release.ENVPREFIX + "_DEST_URL"
			description = "Destination[check] Check Submission URL"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgPort
			longOpt     = "dest-port"
			envVar      = release.ENVPREFIX + "_DEST_PORT"
			description = "Destination[agent|statsd] port (agent=2609, statsd=8125)"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgInstanceID
			longOpt     = "dest-instance-id"
			envVar      = release.ENVPREFIX + "_DEST_INSTANCE_ID"
			description = "Destination[check] Check Instance ID"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgTarget
			longOpt     = "dest-target"
			envVar      = release.ENVPREFIX + "_DEST_TARGET"
			description = "Destination[check] Check target (default hostname)"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgSearchTag
			longOpt     = "dest-tag"
			envVar      = release.ENVPREFIX + "_DEST_TAG"
			description = "Destination[check] Check search tag"
		)

		RootCmd.Flags().String(longOpt, "", desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}
	{
		const (
			key         = config.KeyDestCfgStatsdPrefix
			longOpt     = "dest-statsd-prefix"
			envVar      = release.ENVPREFIX + "_DEST_STATSD_PREFIX"
			description = "Destination[statsd] Prefix prepended to every metric sent to StatsD"
		)

		RootCmd.Flags().String(longOpt, defaults.StatsdPrefix, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.StatsdPrefix)
	}
	{
		const (
			key         = config.KeyDestCfgAgentInterval
			longOpt     = "dest-agent-interval"
			envVar      = release.ENVPREFIX + "_DEST_AGENT_INTERVAL"
			description = "Destination[agent] Interval for metric submission to agent"
		)

		RootCmd.Flags().String(longOpt, defaults.AgentInterval, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.AgentInterval)
	}

	//
	// API
	//
	{
		const (
			key          = config.KeyAPITokenKey
			longOpt      = "api-key"
			defaultValue = ""
			envVar       = release.ENVPREFIX + "_API_KEY"
			description  = "Circonus API Token key or 'cosi' to use COSI config"
		)
		RootCmd.Flags().String(longOpt, defaultValue, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}

	{
		const (
			key         = config.KeyAPITokenApp
			longOpt     = "api-app"
			envVar      = release.ENVPREFIX + "_API_APP"
			description = "Circonus API Token app"
		)

		RootCmd.Flags().String(longOpt, defaults.APIApp, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.APIApp)
	}

	{
		const (
			key         = config.KeyAPIURL
			longOpt     = "api-url"
			envVar      = release.ENVPREFIX + "_API_URL"
			description = "Circonus API URL"
		)

		RootCmd.Flags().String(longOpt, defaults.APIURL, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.APIURL)
	}

	{
		const (
			key          = config.KeyAPICAFile
			longOpt      = "api-ca-file"
			defaultValue = ""
			envVar       = release.ENVPREFIX + "_API_CA_FILE"
			description  = "Circonus API CA certificate file"
		)

		RootCmd.Flags().String(longOpt, defaultValue, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
	}

	// Miscellaneous

	{
		const (
			key         = config.KeyDebug
			longOpt     = "debug"
			shortOpt    = "d"
			envVar      = release.ENVPREFIX + "_DEBUG"
			description = "Enable debug messages"
		)

		RootCmd.Flags().BoolP(longOpt, shortOpt, defaults.Debug, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.Debug)
	}

	{
		const (
			key          = config.KeyDebugCGM
			longOpt      = "debug-cgm"
			defaultValue = false
			envVar       = release.ENVPREFIX + "_DEBUG_CGM"
			description  = "Enable CGM & API debug messages"
		)

		RootCmd.Flags().Bool(longOpt, defaultValue, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaultValue)
	}

	{
		const (
			key          = config.KeyDebugTail
			longOpt      = "debug-tail"
			defaultValue = false
			envVar       = release.ENVPREFIX + "_DEBUG_TAIL"
			description  = "Enable log tailing messages"
		)

		RootCmd.Flags().Bool(longOpt, defaultValue, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaultValue)
	}

	{
		const (
			key          = config.KeyDebugMetric
			longOpt      = "debug-metric"
			defaultValue = false
			envVar       = release.ENVPREFIX + "_DEBUG_METRIC"
			description  = "Enable metric rule evaluation tracing debug messages"
		)

		RootCmd.Flags().Bool(longOpt, defaultValue, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaultValue)
	}

	{
		const (
			key         = config.KeyAppStatPort
			longOpt     = "stat-port"
			envVar      = release.ENVPREFIX + "_STAT_PORT"
			description = "Exposes app stats while running"
		)

		RootCmd.Flags().String(longOpt, defaults.AppStatPort, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.AppStatPort)
	}

	{
		const (
			key         = config.KeyLogLevel
			longOpt     = "log-level"
			envVar      = release.ENVPREFIX + "_LOG_LEVEL"
			description = "Log level [(panic|fatal|error|warn|info|debug|disabled)]"
		)

		RootCmd.Flags().String(longOpt, defaults.LogLevel, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.LogLevel)
	}

	{
		const (
			key         = config.KeyLogPretty
			longOpt     = "log-pretty"
			envVar      = release.ENVPREFIX + "_LOG_PRETTY"
			description = "Output formatted/colored log lines"
		)

		RootCmd.Flags().Bool(longOpt, defaults.LogPretty, desc(description, envVar))
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
		bindEnvError(key, viper.BindEnv(key, envVar))
		viper.SetDefault(key, defaults.LogPretty)
	}

	{
		const (
			key          = config.KeyShowVersion
			longOpt      = "version"
			shortOpt     = "V"
			defaultValue = false
			description  = "Show version and exit"
		)
		RootCmd.Flags().BoolP(longOpt, shortOpt, defaultValue, description)
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
	}

	{
		const (
			key         = config.KeyShowConfig
			longOpt     = "show-config"
			description = "Show config (json|toml|yaml) and exit"
		)

		RootCmd.Flags().String(longOpt, "", description)
		bindFlagError(key, viper.BindPFlag(key, RootCmd.Flags().Lookup(longOpt)))
	}

}

// initConfig reads in config file and ENV variables if set.
func initConfig() {
	if cfgFile != "" {
		viper.SetConfigFile(cfgFile)
	} else {
		viper.AddConfigPath(defaults.EtcPath)
		viper.AddConfigPath(".")
		viper.SetConfigName(release.NAME)
	}

	viper.AutomaticEnv()

	if err := viper.ReadInConfig(); err != nil {
		f := viper.ConfigFileUsed()
		if f != "" {
			log.Fatal().Err(err).Str("config_file", f).Msg("Unable to load config file")
		}
	}
}

// initLogging initializes zerolog.
func initLogging(_ *cobra.Command, _ []string) error {
	//
	// Enable formatted output
	//
	if viper.GetBool(config.KeyLogPretty) {
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stdout})
	}

	//
	// Enable debug logging, if requested
	// otherwise, default to info level and set custom level, if specified
	//
	if viper.GetBool(config.KeyDebug) {
		viper.Set(config.KeyLogLevel, "debug")
		zerolog.SetGlobalLevel(zerolog.DebugLevel)
		return nil
	}

	if viper.IsSet(config.KeyLogLevel) {
		level := viper.GetString(config.KeyLogLevel)

		switch level {
		case "panic":
			zerolog.SetGlobalLevel(zerolog.PanicLevel)
		case "fatal":
			zerolog.SetGlobalLevel(zerolog.FatalLevel)
		case "error":
			zerolog.SetGlobalLevel(zerolog.ErrorLevel)
		case "warn":
			zerolog.SetGlobalLevel(zerolog.WarnLevel)
		case "info":
			zerolog.SetGlobalLevel(zerolog.InfoLevel)
		case "debug":
			zerolog.SetGlobalLevel(zerolog.DebugLevel)
		case "disabled":
			zerolog.SetGlobalLevel(zerolog.Disabled)
		default:
			return fmt.Errorf("Unknown log level (%s)", level) //nolint:goerr113
		}
	}

	return nil
}
